transformer_lens/components.py

```diff
@@ -1332,6 +1336,65 @@ def forward(
         )
 
 
+class MoE(nn.Module):
+    def __init__(self, cfg: Union[Dict, HookedTransformerConfig]):
+        super().__init__()
+        if isinstance(cfg, Dict):
+            cfg = HookedTransformerConfig.from_dict(cfg)
+        self.cfg = cfg
+
+        # Ensure that num_experts and experts_per_token are specified and non-zero
+        assert cfg.num_experts, "num_experts must be specified for MoE layer"
+        assert (
+            cfg.experts_per_token
+        ), "experts_per_token must be specified for MoE layer"
+        assert (
+            cfg.experts_per_token <= cfg.num_experts
+        ), "experts_per_token must be less than or equal to num_experts"
+
+        self.experts = nn.ModuleList(
+            [
+                GatedMLP(cfg) if cfg.gated_mlp else MLP(cfg)
+                for _ in range(cfg.num_experts)
+            ]
+        )
+        self.W_gate = nn.Parameter(
+            torch.empty(cfg.d_model, cfg.num_experts, dtype=cfg.dtype)
+        )
+
+        # Hook on the weights of selected experts [batch pos experts_per_token]
+        self.hook_gate = HookPoint()
```

Comment:
These hook points should perhaps be renamed, the other more verbose option I was considering was `hook_expert_weights` and `hook_expert_indices`, which avoids confusion with the gate hook of gated mlps.